sync/sync.go

```diff
@@ -26,38 +27,84 @@ func Sync(m cluster.Manifests, repoResources map[string]resource.Resource, clus
 	// to figuring out what's changed, and applying that. We're
 	// relying on Kubernetes to decide for each application if it is a
 	// no-op.
-	var sync cluster.SyncDef
+	sync := cluster.SyncDef{}
+
+	nsClusterResources, otherClusterResources := separateResourcesByType(clusterResources)
+	nsRepoResources, otherRepoResources := separateResourcesByType(repoResources)
 
+	// First tackle resources that are not Namespace kind, in case we are deleting the Namespace as well
+	// Deleting a Namespace first, then a resource in this namespace causes an error
+
+	// DANGER ZONE (tamara) This works and is dangerous. At the moment will delete Flux and
+	// other pods unless the relevant manifests are part of the user repo. Needs a lot of thought
+	// before this cleanup cluster feature can be unleashed on the world.
 	if deletes {
-		for id, res := range clusterResources {
-			if res.Policy().Contains(policy.Ignore) {
-				logger.Log("resource", res.ResourceID(), "ignore", "delete")
-				continue
-			}
-			if _, ok := repoResources[id]; !ok {
-				sync.Actions = append(sync.Actions, cluster.SyncAction{
-					ResourceID: id,
-					Delete:     res.Bytes(),
-				})
-			}
+		for id, res := range otherClusterResources {
+			prepareSyncDelete(logger, repoResources, id, res, &sync)
+		}
+		for id, res := range nsClusterResources {
+			prepareSyncDelete(logger, repoResources, id, res, &sync)
 		}
 	}
 
-	for id, res := range repoResources {
-		if res.Policy().Contains(policy.Ignore) {
-			logger.Log("resource", res.ResourceID(), "ignore", "apply")
-			continue
-		}
-		if cres, ok := clusterResources[id]; ok {
-			if cres.Policy().Contains(policy.Ignore) {
-				logger.Log("resource", res.ResourceID(), "ignore", "apply")
-				continue
-			}
+	// To avoid errors due to a non existent namespace if a resource in that namespace is created first,
+	// create Namespace objects first
+	for id, res := range nsRepoResources {
+		prepareSyncApply(logger, clusterResources, id, res, &sync)
+	}
+	for id, res := range otherRepoResources {
+		prepareSyncApply(logger, clusterResources, id, res, &sync)
+	}
+
+	return clus.Sync(sync)
+}
+
+func separateResourcesByType(resources map[string]resource.Resource) (map[string]resource.Resource, map[string]resource.Resource) {
+	if len(resources) == 0 {
+		return nil, nil
+	}
+	nsResources := make(map[string]resource.Resource)
+	otherResources := make(map[string]resource.Resource)
+	for id, res := range resources {
+		_, kind, _ := res.ResourceID().Components()
+		if kind == "namespace" {
```

Comment:
It occurs to me that this test is rather particular to Kubernetes -- if it compared the `Kind` to a constant, that constant would be defined in `cluster/kubernetes`. Indeed, you could argue that the question of what-depends-on-what is particular to the cluster implementation.

But: I will file it as a follow-up issue, since we don't have implementations other than Kubernetes now anyway (i.e., nothing to be done here)

Reply:
You are right. I did not consider other than Kubernetes, I assume because, as you said, we have only deal with Kubernetes at the moment.